consensus/src/dag/round_state.rs

```diff
@@ -0,0 +1,170 @@
+        strong_links: Vec<NodeCertificate>,
+    );
+
+    fn reset(&mut self);
+}
+
+/// Move as fast as 2f+1
+pub struct OptimisticResponsive {
+    event_sender: tokio::sync::mpsc::Sender<Round>,
+}
+
+impl OptimisticResponsive {
+    pub fn new(event_sender: tokio::sync::mpsc::Sender<Round>) -> Self {
+        Self { event_sender }
+    }
+}
+
+#[async_trait]
+impl ResponsiveCheck for OptimisticResponsive {
+    async fn check_for_new_round(
+        &mut self,
+        highest_strong_links_round: Round,
+        _strong_links: Vec<NodeCertificate>,
+    ) {
+        let new_round = highest_strong_links_round + 1;
+        let _ = self.event_sender.send(new_round).await;
+    }
+
+    fn reset(&mut self) {}
+}
+
+/// More sophisticated strategy to move round forward given 2f+1 strong links
+/// Delay if backpressure is triggered. (TODO)
+/// Move as soon as 3f+1 is ready. (TODO: make it configurable)
+/// Move if minimal wait time is reached. (TODO: make it configurable)
+pub struct AdaptiveResponsive {
+    epoch_state: Arc<EpochState>,
+    start_time: Duration,
+    minimal_wait_time: Duration,
+    event_sender: tokio::sync::mpsc::Sender<Round>,
+    event_sent: bool,
+}
+
+impl AdaptiveResponsive {
+    pub fn new(
+        event_sender: tokio::sync::mpsc::Sender<Round>,
+        epoch_state: Arc<EpochState>,
+        minimal_wait_time: Duration,
+    ) -> Self {
+        Self {
+            epoch_state,
+            start_time: duration_since_epoch(),
+            minimal_wait_time,
+            event_sender,
+            event_sent: false,
+        }
+    }
+}
+
+#[async_trait]
+impl ResponsiveCheck for AdaptiveResponsive {
+    async fn check_for_new_round(
+        &mut self,
+        highest_strong_links_round: Round,
+        strong_links: Vec<NodeCertificate>,
+    ) {
+        let new_round = highest_strong_links_round + 1;
+        if self.event_sent {
+            return;
+        }
+        observe_round(
+            self.start_time.as_micros() as u64,
+            RoundStage::StrongLinkReceived,
+        );
+        let voting_power = self
+            .epoch_state
+            .verifier
+            .sum_voting_power(strong_links.iter().map(|cert| cert.metadata().author()))
+            .expect("Unable to sum voting power from strong links");
+        // voting power == 3f+1 or pass minimal wait time
+        let duration_since_start = duration_since_epoch().saturating_sub(self.start_time);
+        if voting_power == self.epoch_state.verifier.total_voting_power()
+            || duration_since_start >= self.minimal_wait_time
+        {
+            let _ = self.event_sender.send(new_round).await;
+            self.event_sent = true;
+        } else {
+            // wait until minimal time reaches before sending
+            let sender = self.event_sender.clone();
+            let wait_time = self.minimal_wait_time.saturating_sub(duration_since_start);
+            tokio::spawn(async move {
```

Comment:
This can be called f times, right? Basically, in every add_node after the first 2f+1 stake is added.
If yes, then new_round will be sent f times and we are wasting tasks, aren't we?

Reply:
I have event_sent boolean to prevent that, but also check in the enter_new_round to deduplicate

Reply:
I see. But now I am back to my previous question here. What if I start a timer but before the timer expires I get all the votes? Then I still have to wait for the timer, right?

Reply:
you see the if branch above right? 🤦

Reply:
Because event_sent is set to true immediately, it prevent  the new added nodes from sending the new_round event.

Reply:
oh I see your point, maybe we should just deal with duplicate events?

Reply:
Maybe I am missing something, but I think it basically means we will always wait for the timeout.

Reply:
I think duplicate events are fine, we can do something fancier like cancelling inflight tasks but probably not worth it

Reply:
Probably. If it is not a performance penalty to spawn all these tasks.

Reply:
How about selecting between two futures? One is a timer and one returns when the voting in the current_round is satisfied. But maybe ugly because it will have to get the DAG.

Reply:
Updated, I think now it's reasonably good

Reply:
nit(I guess): the spawn does not set the state to Sent. This means that the new round event will sent twice. Once from the task and once from the first if which will eventually be satisfied.

Reply:
we abort the task if we send the event directly, in race condition yes we still can receive twice but most cases it should be cancelled properly

Reply:
Yes, I see. When the task sends the event, we will advance the current round and later votes will be simply ignored by the RoundState. So there is only a small period for a possible race. Nice.